[DSpace-CRIS] Porting of authority framework functionalities (Backend)

AdamF42 opened the pull request:
## References

* Fixes #11748
* Requires frontend PR: DSpace/dspace-angular#4956  
* Related to #11549  

## Description
Initial port of the DSpace-CRIS authority framework into core DSpace, enabling real-time entity lookup, confidence-based linking, and automatic graph maintenance through server-side consumers.

## Instructions for Reviewers

The following integration tests document the main behaviours and are the best entry point for reviewers:

  1. Enabling entities by default by using a generic "item" entity
   * IT-Class: `dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemAuthorityIT.java`
   * Methods: `entityTypeNoneAuthorityFilters`
   * Explanation: This test explicitly configures an authority with entityType
     set to "none" (`cris.ItemAuthority.NoneAuthority.entityType=none`) and
     verifies that it correctly matches a generic item (created without an
     explicit entity type) while filtering out items with specific entity types
     like "Person" or "OrgUnit". This addresses the hint "none entity type
     references in codebase".
   * Key Classes:
        - `org.dspace.content.authority.ItemAuthority`: Handles the logic for
            matching items. It checks for the entityType configuration and, if set to
            "none", it bypasses specific type filtering to allow matching generic
            items.
        - `org.dspace.discovery.indexobject.ItemIndexFactoryImpl`: Responsible for
            indexing the dspace.entity.type field into Solr, which allows the discovery
            and authority frameworks to filter items by their entity status (or lack
            thereof).
        - `org.dspace.discovery.configuration.DiscoveryConfiguration`: Used to define
            the `defaultConfiguration` which specifies the Solr filter queries that
            include all items regardless of their specific entity type.
   * Configuration File: `dspace/config/spring/api/discovery.xml`
   * Reference: The `defaultConfiguration` bean defines the
     default search scope.
       * Details: It uses `defaultFilterQueries` that includes
         `search.resourcetype:Item` without restricting to a specific entityType.
         Contrast this with the specific entity configurations. This default configuration enables the
         discovery of generic "items" that do not have a specific
         `dspace.entity.type`.

  2. Improvements to the authority framework (more flexibility by entity type
  and collection)
   * IT-Class: `dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemAuthorityIT.java`
   * Methods: `entityTypeAuthorityFilters`, `authorityNameAuthorityFilters`,
     `personAuthorityTests`, `multiItemAuthorityTest`
   * Explanation: These tests demonstrate the flexibility of the authority
     framework. They show how authorities can be configured to filter
     suggestions based on the `dspace.entity.type` of the target items
     (`solrQuery.addFilterQuery("dspace.entity.type:" + entityType)` in
     `ItemAuthority.java`), allowing different authorities (e.g., `PersonAuthority`) to target specific entity types (e.g., `Person`).
    * Key Classes:
        * `org.dspace.content.authority.ItemAuthority`: The primary class that
            implements the flexible authority search. It dynamically reads the
            entityType from configuration (`cris.ItemAuthority.{authorityName}.entityType`) and
            applies it as a Solr filter query.
        * `org.dspace.content.authority.service.ChoiceAuthorityService`: Orchestrates
            the various authority plugins. It has been improved to handle entity-type-aware lookups and link results back to the submission context.
        * `org.dspace.content.authority.ItemMultiAuthority`: An extension that allows
            an authority to return multiple related items, increasing flexibility for complex entity relationships.
   * Configuration File: `dspace/config/authority.cfg`
   * Reference: `plugin.named.org.dspace.content.authority.ChoiceAuthority`
       * Details: This configuration allows defining multiple named instances of
         `ItemAuthority` (e.g., `PersonAuthority`, `OrgUnitAuthority`). The code
         (`ItemAuthority.java`) then looks for dynamic properties in the format
         `cris.ItemAuthority.{authorityName}.entityType` to filter the authority results by
         `dspace.entity.type`, allowing different authorities to target specific
         entity types.

  3. Improvement to the submission UI
   * IT-Class: `dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemAuthorityIT.java`
   * Method: `singleItemAuthorityTest`
   * Explanation: This test verifies that the authority response contains
     `otherInformation` with keys prefixed with data-. For example, it asserts the
     presence of `data-oairecerif_author_affiliation`. This directly addresses the
     hint "Authorities now handles data- additional entries that can be mapped
     in the FE".
   * Key Classes:
        - `org.dspace.util.ItemAuthorityUtils`: Contains utility methods (like
        buildExtra) used by authorities to package additional metadata into the
        `otherInformation` map of a Choice object.
        - `org.dspace.content.authority.ItemSimpleAuthorityMetadataGenerator` (and other
            ExtraMetadataGenerator implementations): These classes explicitly populate
            the extras map with keys prefixed with data- (e.g.,
            data-oairecerif_author_affiliation), which are then passed to the frontend
            to automatically fill other submission fields.
   * Configuration File: `dspace/config/submission-forms.xml`
   * Reference: <relation-field> and <linked-metadata-field> elements (e.g.,
     inside publicationStep form).
       * Details: The <linked-metadata-field> configuration allows mapping the
         "data-" entries (generated by the backend authority) to specific
         metadata fields in the item. The <externalsources> tag connects the
         field to specific external providers (like ORCID), enabling the UI to
         leverage this rich data.

  4. "Control when linked Items/Entities are created"
   * IT-Class: `dspace-server-webapp/src/test/java/org/dspace/authority/CrisConsumerIT.java`
   * Method: `testItemWithWillBeGeneratedAuthority`
   * Explanation: This test verifies the creation of linked entities using the
     `AuthorityValueService.GENERATE` constant (which resolves to "will be
     generated::"). The `CrisConsumer` class uses this prefix (and "will be
     referenced::") to decide whether to create a new related item or link to an
     existing one.
   * Key Classes:
        * `org.dspace.authority.CrisConsumer`: The core event consumer that listens
            for item installations or modifications. It decides whether to trigger the
            creation of a new related item based on the authority prefix.
        * `org.dspace.authority.service.AuthorityValueService`: Defines the constants
            GENERATE (will be generated::) and REFERENCE (will be referenced::) used to
            signal to the CrisConsumer whether an item should be created or just
            linked.
        * `org.dspace.authority.filler.AuthorityImportFillerService`: Used by the
            CrisConsumer to determine how to populate the newly created item using data
            from the authority source (e.g., pulling researcher details from ORCID).
   * Configuration File: `dspace/config/modules/cris.cfg`
   * Reference:
       * `cris.import.submission.enabled.entity`
       * `cris.import.submission.enabled.entity.use-template`
       * `cris.item-reference-resolution.override-metadata-value`
       * Details: These properties directly control the behavior of the
         CrisConsumer. They determine whether a new entity should be created
         when a "will be generated" authority value is encountered, whether to
         use a template for that creation, and if the local metadata value
         should be overwritten by the authority's value.
5. Multilingual Support for Content and Vocabularies
 * IT-Class: `org.dspace.app.rest.DSpaceRestRepositoryLanguageIT` (Tests metadata filtering and projections).
 * Key Classes:
    * `org.dspace.content.authority.DSpaceControlledVocabulary`: Implements locale-aware initialization. It detects the user's locale and attempts to load corresponding XML files (e.g., srsc_it.xml) to provide translated labels for taxonomy terms.
    * `org.dspace.discovery.SolrServiceMetadataBrowseIndexingPlugin`: Manages the indexing of metadata alongside language tags. This allows the search engine to distinguish between language versions (e.g., using dc.title_en vs dc.title_it logic).
    * `org.dspace.app.rest.projection.AllLanguagesProjection`: Handles the projection=allLanguages request parameter, which instructs the API to return the complete metadata map regardless of the user's current locale.
 * Configuration Files:
    * `dspace/config/dspace.cfg`: Defines webui.supported.locales and default.locale to enable global multi-language support.
    * `dspace/config/modules/authority.cfg`: Configures the behavior of multilingual authority providers.
    * `Vocabulary Files`: XML definitions such as srsc.xml, srsc_it.xml, or countries_de.xml.



### Base Configuration

**NOTE**: to use the authority framework it is necessary to disable the indexing of virtual metadata with the following configuration:
```
relationship.enable-virtual-metadata = false
```

#### Authority Control Settings (`authority.cfg`)
``` 
plugin.named.org.dspace.content.authority.ChoiceAuthority = \
 org.dspace.content.authority.OrcidAuthority = AuthorAuthority

choices.plugin.dc.contributor.author = AuthorAuthority
choices.presentation.dc.contributor.author = suggest
authority.controlled.dc.contributor.author = true

cris.ItemAuthority.AuthorAuthority.entityType = Person

cris.ItemAuthority.AuthorAuthority.source = orcid

orcid.authority.prefix = will be generated::ORCID::
```


## How to test

1. Create a new workspace item with type Publication
2. Search for an author 
<img width="1496" height="324" alt="image" src="https://github.com/user-attachments/assets/15ddf4df-e625-4e39-be58-d8fdc44ad578" />
3. Deposit the item

4. Verify in the search result that the publication has a clickable reference to the author and the popup with author information appear 
<img width="1704" height="1401" alt="image" src="https://github.com/user-attachments/assets/373e3c8d-3a09-452b-ba39-3f8bd83f7b56" />

5. Verify that clicking on the person name, you will be redirected on the Person page
6. Verify that on the Publication page , the authority reference is visible with its popup 
<img width="1704" height="1401" alt="image" src="https://github.com/user-attachments/assets/c4b58825-4a07-4fa8-8605-0df9ccd13f4a" />



### References
[DSpace-CRIS and DSpace Differences](https://wiki.lyrasis.org/display/DSPACE/DSpace-CRIS+and+DSpace+Differences)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

AdamF42 committed:
[DURACOM-427] ported authority framework from dspace-cris

benbosman commented on dspace-api/src/main/java/org/dspace/authority/service/impl/ItemSearcherByMetadata.java:
```diff
@@ -0,0 +1,217 @@
+            return null;
+        }
+
+        IndexableObject indexableObject = indexableObjects.get(0);
```
Only the first result is checked. What if there are multiple matches?

benbosman reviewed (commented):
Thanks for this huge PR.
It will take some time to review and test all of this.
Since this touches entities, it will also be important to test in detail that the normal way of creating and editing entities remains functional.
This is an initial review of some of the code, but not all code, and nothing has been tested yet. Subsequent reviews will follow.

AdamF42 committed:
[DURACOM-427] Add deprecated authority service methods for backward compatibility
Restore 7 deprecated methods delegating to new CRIS authority framework:
- MetadataAuthorityService: 2 methods
- ChoiceAuthorityService: 5 methods
All use Constants.ITEM as default dsoType for Item-centric compatibility.

AdamF42 committed:
[DURACOM-427] Improve error handling in suggestion import for fault tolerance.

Move exception handling from addSuggestion() to PublicationLoader to prevent single suggestion failures from aborting entire import batches.
Individual suggestion errors are now logged while allowing import to continue processing remaining suggestions.

AdamF42 committed:
[DURACOM-427]Fix metadata persistence in AuthorityLinkConsumer

Track modified items during consume() and properly persist changes in end() method by calling itemService.update(). This ensures metadata modifications are saved to the database and items are properly re-indexed.

Also implements memory management by uncaching entities after update to prevent Hibernate memory exhaustion when processing large batches.

Addresses review comment at DSpace/DSpace#11718 (comment r2775092221)

AdamF42 committed:
[DURACOM-427] Fix on AuthorityLinkConsumer

- Exception is no longer silently swallowed
- DCInputsReader inputsReader is now an instance field

Addresses https://github.com/DSpace/DSpace/pull/11718#discussion_r2775168849 and https://github.com/DSpace/DSpace/pull/11718#discussion_r2775179699

AdamF42 committed:
[DURACOM-427] Rename performSearchByMetadata to findFirstByMetadata for clarity

Rename method to explicitly indicate that only the first matching item is returned when multiple results exist. This makes the intentional behavior clear to callers.

Addresses review comment at DSpace/DSpace#11718 (comment r2775211479)

AdamF42 commented on dspace-api/src/main/java/org/dspace/authority/service/impl/ItemSearcherByMetadata.java:
```diff
@@ -0,0 +1,217 @@
+        }
+    }
+
+    private void resolveReferences(Context context, List<MetadataValue> metadataValues, Item item)
```
Thanks for raising this performance concern. You're correct that this operation can be
expensive in certain scenarios. Let me explain the design rationale and current mitigations:

**What this code does:**
When a Person (or other entity) with an identifier (ORCID, RID, ISSN, DOI, etc.) is created or 
updated, the `resolveReferences` method searches the entire repository for items with **placeholder 
authority references** (e.g., `"will be referenced::ORCID::0000-0002-1825-0097"`) and updates them 
to use the actual item UUID.

**Current performance optimizations:**
1. **ThreadLocal caching** (lines 69-71): Prevents duplicate lookups within the same consumer execution
2. **Deduplication** (`ItemReferenceResolverConsumer:41`): Prevents processing the same item multiple times in one batch
3. **Selective execution**: Only runs when authority-controlled identifier fields change

**Performance characteristics:**
- For typical use cases (person with <100 publications): **Acceptable performance**
- For high-profile entities (person with 1000+ publications): **Could be slower, but manageable**

**For this PR:**
Given that this implementation is already tested in production and works well for the typical use cases, I propose we proceed with the current implementation as-is. If specific performance bottlenecks are identified in the future, we can open a follow-up issue to explore optimizations

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/filler/AuthorityImportFiller.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param context        the context
+     * @param relatedItem    the source metadata
+     * @param metadata       the source metadata related to the list to return
```
I'm not understanding the JavaDocs here.  What is the list of metadata returned? Is it coming from the "relatedItem", or the "metadata" param or both?  Because the methods that override this interface have no JavaDocs of there own, it seems important to provide clearer details here.

Also the param descriptions here seem confusing or wrong.  The `relatedItem` is described as "source metadata"?  Is that correct?

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/filler/AuthorityImportFillerService.java:
```diff
@@ -0,0 +1,18 @@
+
+    String SOURCE_INTERNAL = "INTERNAL-SUBMISSION";
+
+    AuthorityImportFiller getAuthorityImportFillerByMetadata(MetadataValue metadata);
```
Please add JavaDocs to this method as it doesn't appear to be a simple "getter".  It's unclear how MetadataValue maps to `AuthorityImportFiller` here.

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/filler/ExternalDataProviderImportFiller.java:
```diff
@@ -0,0 +1,140 @@
+
+/**
+ * Implementation of {@link AuthorityImportFiller} which enriches the item with
+ * the data retrieved with the configured
```
This description is incomplete (the sentence abruptly ends). Please provide a more detailed description of this class in the JavaDocs here.

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/service/ItemReferenceResolver.java:
```diff
@@ -0,0 +1,34 @@
+public interface ItemReferenceResolver {
+
+    /**
+     * Resolve all the references to the given one via will be referenced authority.
```
I don't understand this sentence.   Is this referencing the "will be referenced" values?  What does "resolving references" mean?  Is it filling out the "will be referenced" values with the actual references?  I think it'd be good to explain what "resolve all the references" means here.

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/service/ItemSearcherMapper.java:
```diff
@@ -0,0 +1,42 @@
+import org.dspace.core.Context;
+import org.springframework.util.Assert;
+
+public class ItemSearcherMapper {
```
This class needs JavaDocs for the class and all public methods

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/AuthorityLinkConsumer.java:
```diff
@@ -0,0 +1,163 @@
+ * @author Stefano Maffei(stefano.maffei at 4science.com)
+ *
+ */
+public class AuthorityLinkConsumer implements Consumer {
```
This Consumer is missing a JavaDoc description

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/AuthorityValue.java:
```diff
@@ -143,6 +147,30 @@ public void delete() {
+        this.serviceId = serviceId;
+    }
+
+    public Map<String, List<String>> getOtherMetadata() {
```
In either this method or `addOtherMetadata` could we describe what "other metadata" means?  What's an example where we need more metadata for the AuthorityValue?

tdonohue commented on dspace-api/src/main/java/org/dspace/authority/CrisConsumer.java:
```diff
@@ -0,0 +1,346 @@
+
+/**
+ * Consumer to store item related entities when an item submission/modification
+ * occurs.
```
Because there's no more JavaDocs or inline comments below, I'd appreciate it if we can provide more details in the JavaDocs here.  This consumer should better describe what it is storing and doing to update the item (or give examples)

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/AuthorityNameAuthorityFilter.java:
```diff
@@ -0,0 +1,31 @@
+
+import org.apache.commons.collections.CollectionUtils;
+
+public class AuthorityNameAuthorityFilter extends CustomAuthorityFilter {
```
Please add JavaDocs to describe this new class

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/CustomAuthorityFilter.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @return a list of custom solr filter queries
+     */
+    public List<String> getFilterQueries(LinkableEntityAuthority linkableEntityAuthority) {
```
Please update JavaDocs to include a description of what the method does. That's missing.

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/CustomAuthoritySolrFilter.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author Stefano Maffei 4Science.com
+ */
+public interface CustomAuthoritySolrFilter {
```
Please update JavaDocs to add a description of this interface

tdonohue replied:
@AdamF42 : Thanks for all the updated JavaDocs you've added already! It all looks good except that you overlooked updating this `CustomAuthoritySolrFilter` interface with JavaDocs (at the top of the file).

tdonohue reviewed (changes requested):
@AdamF42 : I'm working my way through a code review.  I'm only about 1/3 of the way through, but I'm finding a **lot** of missing or incorrect JavaDocs.  We need to ensure all this new code is coming with JavaDocs (and tests where possible), as otherwise it will be difficult for developers to support it.

See below a lot of references (inline) to missing/incorrect JavaDocs.  The code actually looks good to me so far, it's just all missing JavaDocs.   I'll continue with more reviewing tomorrow, but keep in mind I'll continue combing through for missing JavaDocs.

MarieVerdonck commented on dspace-api/src/main/java/org/dspace/authority/CrisConsumer.java:
```diff
@@ -0,0 +1,346 @@
+            }
+
+            if (!relatedItemAlreadyPresent) {
+                Collection collection = collectionService.retrieveCollectionByEntityType(context, item, entityType);
```
This just retrieves the first collection with entity type "Person". There may be multiple in a repository, some may be used for specific purposes. Better if this collection retrieval is based on a configured UUID, and optionally if configured UUID is not set, it can fallback on _if_ there is just one collection with entity Person use that. If there are multiple collections with entity Person I would at least log a warning.

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/ItemAuthorityExtraMetadataGenerator.java:
```diff
@@ -0,0 +1,28 @@
+
+    Map<String, String> build(String authorityName, SolrDocument solrDocument);
+
+    List<Choice> buildAggregate(String authorityName, SolrDocument solrDocument);
```
Both methods in this interface need JavaDocs

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/PersonCoarseCustomSolrFilterImpl.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @author Stefano Maffei 4Science.com
+ */
+public class PersonCoarseCustomSolrFilterImpl extends ItemAuthorityCustomSolrFilterImpl {
```
Same comments here. JavaDocs is missing for this class & the Solr queries below could use a comment to describe the query

tdonohue commented on dspace-api/src/main/java/org/dspace/content/authority/PersonStrictCustomSolrFilterImpl.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author Stefano Maffei 4Science.com
+ */
+public class PersonStrictCustomSolrFilterImpl implements CustomAuthoritySolrFilter {
```
JavaDocs is missing for this class

tdonohue commented on dspace-api/src/main/java/org/dspace/content/enhancer/impl/RelatedEntityItemEnhancer.java:
```diff
@@ -0,0 +1,341 @@
+
+/**
+ * Implementation of {@link ItemEnhancer} that add metadata values on the given
+ * item taking informations from linked entities.
```
As this is a large class with no inline comments, could you add more details about *how* this enhances the given item?  Either that, or provide JavaDocs on the `enhance()` method below to describe what it's doing.

tdonohue commented on dspace-api/src/main/java/org/dspace/content/service/CollectionService.java:
```diff
@@ -494,6 +510,29 @@ public int countCollectionsWithSubmit(Context context, String q, Community commu
+     * NOTE: for better performance, this method retrieves its results from an index (cache)
+     * and does not query the database directly.
+     * This means that results may be stale or outdated until
+     * https://github.com/DSpace/DSpace/issues/2853 is resolved."
```
NOTE: The referenced issue ticket was recently resolved, so we probably can remove this last sentence.

tdonohue commented on dspace-api/src/main/java/org/dspace/core/Context.java:
```diff
@@ -475,6 +477,14 @@ public void commit() throws SQLException {
         }
     }
 
+    public void clear() {
```
Please add Javadocs to this new public method.  Make sure to describe how this is different from `rollback()` and `close()` and `abort()`.

tdonohue commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceStrictBestMatchIndexingPlugin.java:
```diff
@@ -0,0 +1,120 @@
+
+    private static final String EXCLUDE_NUMBERS_CONFIG = "solr-service.strict-best-match.exclude.numbers";
+
+    private static final String NORMALIZE_WHITESPACES = "solr-service.strict-best-match.exclude.normalize-whitespaces";
```
All of the above constants appear to be configurations, but I don't see them in a `*.cfg` file.  We should add them to a configuration file and describe their behavior.

tdonohue commented on dspace/config/registries/cris-types.xml:
```diff
@@ -0,0 +1,306 @@
+        <schema>cris</schema>
+        <element>harvesting</element>
+        <qualifier>email</qualifier>
+        <scope_note></scope_note>
```
If possible, could we add a `<scope_note>` to all the new metadata fields (not just in this file but in all the files)?  I think it'd be helpful to others to understand what each field is used for if we can add a scope note to each.

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -791,12 +791,22 @@ aip.disseminate.dmd = MODS, DIM
 # Add iiif here, if you are using dspace-iiif.
 # Add orcidqueue here, if the integration with ORCID is configured and wish to enable the synchronization queue functionality
-event.dispatcher.default.consumers = versioning, discovery, eperson, qaeventsdelete, ldnmessage, customurl, audit
+event.dispatcher.default.consumers = versioning, discovery, eperson, crisconsumer, orcidqueue, audit, qaeventsdelete, referenceresolver,  itemenhancer, customurl, iiif, authoritylink, ldnmessage
```
Is this supposed to have `itemenhancer` added?  Why do we need the separate dispatcher if both have the same list of consumers?

(It's possible this question would be answered if you add more details about the `RelatedItemEnhancerUpdatePoller` in general, like I described above)

tdonohue reviewed (changes requested):
@AdamF42 : Thanks again for this PR.  I've completed my initial code review.  All the code looks good overall, but I've found a **lot of missing JavaDocs**.  So, apologies for the very long list of comments (over 60 here and 20+ previously), but we need to ensure that all code is meeting DSpace best practices and that includes JavaDocs.

There are also some minor suggestions/questions inline as well in some areas of the code. If any of these need to involve detailed follow-up work, we can always move them to a bug ticket and address them later.  Thanks again!

AdamF42 committed:
[DURACOM-427] Updated JavaDoc and removed org.dspace.authority.filler.MetadataConfiguration.MappingDetails#visibility

AdamF42 committed:
[DURACOM-427] Updated JavaDoc

nwoodward reviewed (commented):
Thanks for this PR @AdamF42! I am still testing, and I will have more to say later. 

It took me a minute to find `item.enable-virtual-metadata` in the code and where it is set in `./dspace-api/src/test/data/dspaceFolder/config/local.cfg`. Should it be added to `dspace/config/local.EXAMPLE` and commented out?

vins01-4science committed:
docs: places missing JavaDocs for comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835348394

ref: DURACOM-427

vins01-4science committed:
docs: places missing JavaDocs for comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835340190

ref: DURACOM-427

vins01-4science committed:
docs: places missing JavaDocs for comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835330530

ref: DURACOM-427

vins01-4science committed:
fix: removes discovery metadata for comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835325677

ref: DURACOM-427

vins01-4science committed:
doc: adds title description in cris-types.xml resolves comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835293723

ref: DURACOM-427

vins01-4science committed:
fix: Removes unnecessary miur-types and related configuration resolves comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835291024

ref: DURACOM-427

vins01-4science committed:
fix: Removes old metadata from dublin-core-types.xml resolves comment: https://github.com/DSpace/DSpace/pull/11718#discussion_r2835266780

ref: DURACOM-427

vins01-4science committed:
fix: Renames cris-types.xml namespace

ref: DURACOM-427

AdamF42 commented:
> **Clutter linked metadata**
> 
> The item with linked authors like this have a lot of clutter metadata in e.g. `cris.virtual.department`. All containing `#PLACEHOLDER_PARENT_METADATA_VALUE#"` => If they don't have any added value info to the item, they should not be included. Show up in rest endpoints, full item page, etc, even for non admins, but just making admin-only seems insufficient, since would then also clutter admin views and exports...
> 
> Clutter linked md (with value #PLACEHOLDER_PARENT_METADATA_VALUE#) should not be added on the Publication item. I presume this `cris.virtual.*` and `crist.virtualsource.*` md is intended to add info on the linked Person on the Publication item (similar to entities' virtual metadata).
> 
> In the below example I would expect `cris.virtual.orcid` to be populated with the orcids of the linked Person items. They have `cris.sourceid:ORCID::{Orcid}`
> 
> It would be better if `cris.virtual.*` is only populated _if_ there is useful data to put within it, and that `cris.virtualsource.*` is not used, but rather that you put the authority id (uuid of linked person) in the authority field of the corresponding `cris.virtual.*` <img alt="Screenshot 2026-02-20 at 13 08 04" width="708" height="670" src="https://private-user-images.githubusercontent.com/5233724/552684071-f95dcca6-5704-4909-8427-1e66c1a727ec.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzIwMzgxNjYsIm5iZiI6MTc3MjAzNzg2NiwicGF0aCI6Ii81MjMzNzI0LzU1MjY4NDA3MS1mOTVkY2NhNi01NzA0LTQ5MDktODQyNy0xZTY2YzFhNzI3ZWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjVUMTY0NDI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yzk1NDBmYTJhNGU5NzQwYzdlZWI0ZjA5MTM0NjI1ZWI4MDI5OTBmODU1ZGJjODg1Y2QwZDQwNmU1ZDRjNzZlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.1Y28-f6Z_0jjzEzORKNtMa6YfYluf9JfCG4ssO4cbnQ">


@MarieVerdonck, thank you for the feedback.

The use of `#PLACEHOLDER_PARENT_METADATA_VALUE#` is a design choice to enable the `ItemEnhancerService` to resolve values from linked items using the authority at a later stage. In the example you provided, the virtual metadata is generated based on [this](https://github.com/DSpace/DSpace/blob/8deee672861593eb24914705d309c94d48397d5d/dspace/config/spring/api/metadata-enhancers.xml#L34) configuration.

**How it works:**
- `cris.virtualsource.orcid` and `cris.virtualsource.department`: These store the authority (UUID) of the related `dc.contributor.author` or `dc.contributor.editor` used to enrich the publication.
- `cris.virtual.orcid` and `cris.virtual.department`: These contain the `#PLACEHOLDER_PARENT_METADATA_VALUE#` when the system is not yet able to resolve the `person.identifier.orcid` or `person.affiliation.name` from the related authority.

**Why placeholders are used:**
The placeholder is used to store values that can potentially be resolved in the future. In your example, the first author with an authority is `ab603993-3126-44cd-9540-943610e09e53`. Since this item does not currently contain `person.affiliation.name` or `person.identifier.orcid`, both `cris.virtual.department` and `cris.virtual.orcid` will contain the placeholder. The same behavior applies to `85037466-bfb2-493f-bc90-020d5b90696a` and `7366cbdf-2e44-41d0-98f4-9acabf1df5af`.

We create a `cris.virtual.<qualifier>` and `cris.virtualsource.<qualifier>` for every valid authority configured in the [`metadata-enhancers.xml`](https://github.com/DSpace/DSpace/blob/24559748528cf899870c233e5b036b26a1d5c21b/dspace/config/spring/api/metadata-enhancers.xml#L34). Every `cris.virtual.<qualifier>` is related to its `cris.virtualsource.<qualifier>` counterpart **by position**.

It allows the [`RelatedItemEnhancerUpdatePoller`](https://github.com/DSpace/DSpace/blob/24559748528cf899870c233e5b036b26a1d5c21b/dspace-server-webapp/src/main/java/org/dspace/app/rest/enhancer/RelatedItemEnhancerUpdatePoller.java#L35) to process these values asynchronously and ensures the [`ItemEnhancerServiceImpl`](https://github.com/DSpace/DSpace/blob/24559748528cf899870c233e5b036b26a1d5c21b/dspace-api/src/main/java/org/dspace/content/enhancer/service/impl/ItemEnhancerServiceImpl.java#L30) knows exactly which linked item needs to be checked to "resolve" the placeholder into a real value. Take a look at [`ItemEnhancerConsumerIT`](https://github.com/4Science/DSpace/blob/24559748528cf899870c233e5b036b26a1d5c21b/dspace-api/src/test/java/org/dspace/content/enhancer/consumer/ItemEnhancerConsumerIT.java) to have a better understanding.

Currently, it is not possible to avoid using `cris.virtualsource.<qualifier>`, but we can consider hiding them in the frontend.

AdamF42 committed:
[DURACOM-427] Updated JavaDoc

AdamF42 committed:
[DURACOM-427] Fix to map orcid extra metadata.

tdonohue reviewed (commented):
@AdamF42 and @vins01-4science : Based on the discussions in today's Developers meeting around moving some metadata fields around (e.g. from `cris.*` to `dspace.*`), I realized we may want to consider that for some of the other new metadata fields added in this PR.

I've called out below where we might move some fields around.  However, if you feel any of this is **too complex/difficult** for this PR, we could create a separate issue ticket to do this work later on.  I just wanted to call out that we have some duplicative schema.

AdamF42 committed:
[DURACOM-427] Updated JavaDoc

AdamF42 committed:
[DURACOM-427] Fix on multilanguage vocabulary and value-pair

AdamF42 committed:
[DURACOM-427] Updated JavaDoc

tdonohue commented on dspace-server-webapp/src/test/data/dspaceFolder/config/spring/api/test-discovery.xml:
```diff
@@ -1275,6 +1291,115 @@
         <property name="exposeMinAndMaxValue" value="true"/>
     </bean>
 
+<!--    &lt;!&ndash; Override default site-wide configuration to include multilanguage facets &ndash;&gt;-->
```
@AdamF42 : There's a larger block of commented out tests configurations here. This should either be removed or uncommented.

tdonohue commented on dspace/config/spring/api/metadata-enhancers.xml:
```diff
@@ -0,0 +1,66 @@
+>
+
+    <context:annotation-config />
+    
```
Same with this Spring config file. It'd be useful to describe it's purpose in a comment near the top of the file.  Or, alternatively, add a brief comment above each of the configurations below to describe the default settings that are added here.

tdonohue commented on dspace/config/modules/authority.cfg:
```diff
@@ -86,3 +87,233 @@ org.dspace.content.authority.EPersonAuthority = EPersonAuthority
+# SolrSuggestAuthority.dc_subject.facetname = subject
+
+
+# DSpace-CRIS stores by default the authority of controlled vocabularies
```
As noted previously: Please remove the name "DSpace-CRIS" from this line. Instead, describe what this configuration does when set to true or false. For example, where is the authority of controlled vocabularies stored when set to "true". Is there any use case for setting it to "false"?

tdonohue reviewed (changes requested):
@AdamF42 and @vins01-4science : Thanks for all the recently added JavaDocs!! I've verified that *most* of my prior feedback is resolved (though I've called out a few places where feedback has been overlooked below).  

I've added a few more minor comments to this review inline below.  I'm also still waiting on response regarding [my questions above about possibly moving some fields between metadata schemas](https://github.com/DSpace/DSpace/pull/11718#pullrequestreview-3862255636) (if this is not easily possible we could create a ticket for it).

Finally, I tested this today with the frontend PR, and it's working well overall. But, I've found a backend bug with the Author field when using the recommended configs:
```
item.enable-virtual-metadata = false
plugin.named.org.dspace.content.authority.ChoiceAuthority = \
 org.dspace.content.authority.OrcidAuthority = AuthorAuthority

cris.ItemAuthority.AuthorAuthority.entityType = Person
cris.ItemAuthority.AuthorAuthority.source = orcid

choices.plugin.dc.contributor.author = AuthorAuthority
choices.presentation.dc.contributor.author = suggest
authority.controlled.dc.contributor.author = true
orcid.authority.prefix = will be generated::ORCID::
```
Here's how to reproduce it:
1. Start the backend and frontend PR. I'm running the frontend in production mode (npm run build:prod && npm run serve:ssr), but I don't think that matters.
2. Start a submission for a new Publication entity
3. In the Author field, attempt to lookup an author.  Some searches will work (e.g. "Donohue" or "Smith" work for me).  But, some searches **will throw a backend error**.  Some examples are the searches "Adam" or "Long" both will throw this error in the backend logs (this example shows me searching "adam" as you can see from the Solr request):
```
2026-03-04 21:02:00,802 INFO  04822c4e-48e1-4717-b332-1346962b6137 2d4e321e-4a87-4a77-b5a8-dc2e452e8ecb org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Before request [GET /server/api/submission/vocabularies/AuthorAuthority/entries] originated from /workspaceitems/3349/edit
2026-03-04 21:02:00.813 INFO  (qtp6422064-23-null-541) [c: s: r: x:search t:null-541] o.a.s.c.S.Request webapp=/solr path=/select params={q={!lucene+q.op%3DAND+df%3Ditemauthoritylookup}(adam*)^100+OR+("adam")^10&start=0&fq=search.resourcetype:Item&fq=withdrawn:false&fq=NOT(discoverable:false)&fq=dspace.entity.type:Person&rows=20&wt=javabin&version=2} hits=0 status=0 QTime=3
2026-03-04 21:02:02,573 ERROR 04822c4e-48e1-4717-b332-1346962b6137 2d4e321e-4a87-4a77-b5a8-dc2e452e8ecb org.dspace.content.authority.OrcidAuthority @ An error occurs performing profiles search on ORCID registry
java.lang.NullPointerException: Cannot invoke "String.matches(String)" because "familyName" is null
   at org.dspace.content.authority.OrcidAuthority.getTitle(OrcidAuthority.java:183)
   at org.dspace.content.authority.OrcidAuthority.convertToChoice(OrcidAuthority.java:170)
   at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
   at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
   at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
   at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
   at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
   at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
   at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
   at org.dspace.content.authority.OrcidAuthority.performOrcidSearch(OrcidAuthority.java:140)
   at org.dspace.content.authority.OrcidAuthority.getMatches(OrcidAuthority.java:117)
   at org.dspace.app.rest.repository.VocabularyEntryLinkRepository.filter(VocabularyEntryLinkRepository.java:88)
```

It's possible this is an error of invalid data in my database.  But, it still seems like something we could handle more gracefully, instead of throwing a large exception. 

Beyond that, everything else looks to be working.  

As I'm sure you're both aware, this is blocking several other merger PRs. I'd really like to see this PR (and the frontend) merged soon to unblock everything else.  So, if you are able to address or respond to this feedback quickly, I'd love to give it updated testing/review.

AdamF42 committed:
[DURACOM-427] Updated configs descriptions

AdamF42 committed:
[DURACOM-427] Fix OrcidAuthority for possible null givenName and familyName

AdamF42 committed:
[DURACOM-427] removed duplicated code

tdonohue reviewed (commented):
@AdamF42 and @vins01-4science and @atarix83 :  First off, thanks for the updates.  I've verified the [prior bug](https://github.com/DSpace/DSpace/pull/11718#pullrequestreview-3890968028) that I've reported is fixed.  I'm getting close to a +1 on this PR, as the functionality works well. But, I still see some outstanding prior feedback that has not been responded to or addressed:

1.  I asked for additional comments to be added for all new configs in the `authority.cfg`.  These new configs have no details on how to use them.  Two prior comments relate to this request: https://github.com/DSpace/DSpace/pull/11718/changes#r2886347282 and https://github.com/DSpace/DSpace/pull/11718/changes#r2886348596  (If these links don't work, just open the `authority.cfg` in the "Files Changed" section and you'll see my two unresolved comments.  They also were in [my prior review](https://github.com/DSpace/DSpace/pull/11718#pullrequestreview-3890968028))
2. I still really don't like the `if` statement to check for H2 versus PostgreSQL in the `ItemForMetadataEnhancementUpdateDAOImpl` (see comment inline below again).  If we need to keep it for now, I'll move this to a separate bug ticket.
3. Regarding the "cris.*" schema and the various "cris[entity].*" schemas.  I'm OK with @atarix83 's approach [in his last comment](https://github.com/DSpace/DSpace/pull/11718#issuecomment-4004438769).  We can let them come into DSpace as-is, but should deprecate the "cris.*" schema.  The other schemas we can move to future discussions (if needed).

Finally, after some more thought today, I feel we should rename the new `item.enable-virtual-metadata` setting to be `relationship.enable-virtual-metadata` and add it to the existing `relationship.cfg` file (as this configuration impacts that existing entity relationship feature).  See comment inline below.

Other than that, this PR is looking good to me.  I'm not finding any other bugs or anything. So, I'm nearly a +1. 

(I'd also encourage other reviewers/testers like @MarieVerdonck , @benbosman or @nwoodward  to provide updated feedback.  I'd really like to see this PR get moved forward soon, so that we can unblock the other DSpace-CRIS PRs that depend on it.)

AdamF42 committed:
[DURACOM-427] Fix missing SolrServiceValuePairsIndexPlugin in discovery.xml

AdamF42 committed:
[DURACOM-427] Retrieved nodeId from auth to get vocabulary label

AdamF42 committed:
[DURACOM-427] Fixed auth indexing with multilanguage

AdamF42 committed:
[DURACOM-427] Updated missing crisrp types

AdamF42 committed:
[DURACOM-427] Updated authority.cfg config property description

AdamF42 committed:
[DURACOM-427] renamed item.enable-virtual-metadata to relationship.enable-virtual-metadata and updated description

AdamF42 committed:
[DURACOM-427] moved cris schema to dspace one

nwoodward commented:
I'm testing this PR again today, and I see mention of setting the parameter `orcid.authority.prefix = will be generated::ORCID::` in the initial suggested configuration and in recent comments on the PR, but I don't see it in `authority.cfg` or elsewhere. Does it need to be added to the configuration?

AdamF42 committed:
[DURACOM-427] updated authority.cfg with orcid extra configs

vins01-4science commented:
> For [#11718 (comment)](https://github.com/DSpace/DSpace/pull/11718#issuecomment-3967403391)
> 
> => The Person is not being populated with the `person.*` fields like in your example, so no `person.identifier.orcid`, only orcid in `dspace.sourceId` with prefix
> 
> Debugging CrisConsumer#fillRelatedItem =>
> 
> ```
> fillItem:92, ExternalDataProviderImportFiller (org.dspace.authority.filler)
> fillRelatedItem:356, CrisConsumer (org.dspace.authority)
> consumeItem:206, CrisConsumer (org.dspace.authority)
> consume:156, CrisConsumer (org.dspace.authority)
> ```

Hello @MarieVerdonck , thank you for this feedback.

You need to have an account to access these resources.
In the near future most of these "free" api would shift in that way in order to offer a better service to a wider range of institutions, by avoiding interferences with requests made by bots or automated system.
In this PR we're not going to change this behavior. You can open a follow-up ticket, but I think that the right approach is to use the API keys to search for users and then parse the response accordingly.

---

> For https://github.com/DSpace/DSpace/pull/11718#issuecomment-3993904433

Fragmented facets a browse values with suggested MultiLanguageDiscoverSearchFilterFacet OK, though actually requires rename from "subject" to "subjects" and not just changing it's type, unclear why.

Display of the value in UI everywhere does not change depending on the user's selected language though. Would expect this to be translated in:

    Browse: in clickable results & in header of resulting filtered page once clicked
    Facet: filters sidebar values & once selected in filter displayed
    In submission (and hopefully thus making sure same subject is not added twice, under separate languages)


Please share your configuration or at least push it to a public repo in order to have a look on that so we can suggest the right configuration and find out what is missing in your config and what can be marked as a regression compared to the CRIS codebase.

MarieVerdonck reviewed (approved):
For the ORCID info not being added => will retest once I have Member API creds, and create issue if not ok.

For multi-lingual vocabulary controlled values not being translated everywhere created improvement request issue: https://github.com/DSpace/DSpace/issues/12115

Sidenote: some documentation & testing cfg files still contain `item.enable-virtual-metadata` => since it's renamed to `relationship.enable-virtual-metadata` => would be good to adjust everywhere.

AdamF42 committed:
[DURACOM-427] Update to rename old item.enable-virtual-metadata to relationship.enable-virtual-metadata

AdamF42 committed:
[DURACOM-427] Update to remove unused variables

tdonohue reviewed (approved):
👍 Thanks @AdamF42 and @vins01-4science !  I've verified that all my feedback has been addressed **except** for the one issue we've decided to move to a bug ticket.
* I've moved my concerns with `ItemForMetadataEnhancementUpdateDAOImpl` to a blocker ticket in #12119
* I've also decided to create a separate ticket for documenting the features that require `relationship.enable-virtual-metadata` to be set to true versus false, as that's still unclear as of yet.  See #12120

I've also retested everything today and verified the following:
* By default, DSpace acts as normal, and I can successfully create new Entities and relate them via existing DSpace relationships
* But, if I set `relationship.enable-virtual-metadata=false` (and add the other configs in this PR description), I'm then able to create Entities with DSpace-CRIS style relationships

Therefore, I feel this PR looks good enough to merge.  If there are any outstanding issues from anyone else, please move those to a bug ticket and we'll ensure it is completed in a future PR.

tdonohue commented:
@AdamF42 and @vins01-4science : I forgot to mention, I'm adding the `needs documentation` label to the PR because we need to add documentation related this this PR to our DSpace 10 documentation: https://wiki.lyrasis.org/display/DSDOC10x/
